src/dag/chunk.rs

```diff
@@ -1,8 +1,100 @@
-pub type Key = String;
+use super::meta_generated;
+use flatbuffers;
+use flatbuffers::FlatBufferBuilder;
 
 #[allow(dead_code)]
+#[derive(Debug)]
 pub struct Chunk {
-    pub key: Key,
-    buf: Vec<u8>,
-    refs: [Key],
+    hash: String,
+    data: Vec<u8>,
+    meta: Option<(Vec<u8>, usize)>,
+}
+
+#[allow(dead_code)]
+impl Chunk {
+    pub fn new(hash: String, data: Vec<u8>, refs: &[&str]) -> Chunk {
+        Chunk {
+            hash,
+            data,
+            meta: Chunk::create_meta(refs),
+        }
+    }
+
+    pub fn read(hash: String, data: Vec<u8>, meta: Option<Vec<u8>>) -> Chunk {
+        Chunk {
+            hash,
+            data,
+            meta: meta.map(|v| (v, 0)),
+        }
+    }
+
+    pub fn hash(&self) -> &str {
+        &self.hash
+    }
+
+    pub fn data(&self) -> &[u8] {
+        &self.data
+    }
+
+    pub fn refs<'a>(&'a self) -> Box<dyn Iterator<Item = &str> + 'a> {
+        match self.meta() {
```

Comment:
It seems like you ought to be able to say something like:

```
meta_generated::meta::get_root_as_meta(self.meta()?).refs()?.or(std::iter::empty())
```

... but could not figure that out.

Reply:
I can barely read this, so I don't know if I'd favor the above!

`?`for `Options`s only works if the return type of the function is itself an `Option`, so not applicable here.

Reply:
Yeah I just mean that conceptually it would be nice if you could use the question mark iterator on a chain to get an Option, then at the end convert that to a Result.